snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/TrackerControllerImpl.kt

```diff
@@ -70,6 +72,50 @@ class TrackerControllerImpl  // Constructors
         return tracker.track(event)
     }
 
+
+    override fun decorateLink(
+        uri: Uri,
+        extendedParameters: List<CrossDeviceParameter>,
+        reason: String
+    ): Uri? {
+        // UserId is a required parameter of `_sp`
+        if (this.session?.userId == null) {
+            Logger.track(TAG, "$uri could not be decorated as session.userId is null ")
+            return null
+        }
+
+        val values = hashMapOf(
+            CrossDeviceParameter.SESSION_ID to (this.session?.sessionId ?: ""),
+            CrossDeviceParameter.SOURCE_ID to base64Encode(this.appId),
+            CrossDeviceParameter.SOURCE_PLATFORM to this.devicePlatform.value,
+            CrossDeviceParameter.SUBJECT_USER_ID to base64Encode(this.subject.userId ?: "")
```

Comment:
I think this `Util.base64Encode` function does not produce the URL-safe variant of base 64 encoding. To get the URL safe version (as done in the JS tracker), I think you will need to pass these options to the [Android Base64](https://developer.android.com/reference/android/util/Base64) class:

```kt
Base64.encodeToString(string.toByteArray(), Base64.URL_SAFE | Base64.NO_PADDING | Base64.NO_WRAP)
```